caffe2/core/c10_operator.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include <vector>
+#include <ATen/core/dispatch/OpSchemaRegistration.h>
+#include <ATen/core/dispatch/KernelRegistration.h>
+#include <ATen/core/function_schema.h>
+
+namespace caffe2 {
+namespace detail {
+
+template<class Caffe2Operator> const c10::OperatorHandle& c10_op_handle_for_c2_op();
+template <class Caffe2Operator>
+void call_caffe2_op_from_c10(c10::Stack* stack, c10::KernelCache* cache) { // TODO Pass in correct cache type
+  // precondition: on the stack, there's an IValue for each input and an IValue for each output.
+  // The output ones could either be a preallocated tensor or ivalue::None.
+
+  const auto& schema = c10_op_handle_for_c2_op<Caffe2Operator>().schema();
+  const size_t num_outputs = schema.returns().size();
+  const size_t total_num_arguments = schema.arguments().size();
+  const size_t num_inputs = total_num_arguments - num_outputs;
+
+  c10::ArrayRef<c10::IValue> inputs = torch::jit::peekSlice(*stack, 0, num_inputs, total_num_arguments);
+  c10::ArrayRef<c10::IValue> outputs = torch::jit::peekSlice(*stack, num_inputs, num_outputs, total_num_arguments);
+
+  auto inputsVec = inputs.vec();
+  auto outputsVec_ = outputs.vec();
+
+  const auto device = at::Device(at::DeviceType::CPU); // TODO Handle GPU devices
+
+  for (auto& output : outputsVec_) {
+    if (output.isNone() || (output.isTensor() && !output.toTensor().defined())) {
+      output = at::Tensor(c10::C10Tensor(caffe2::empty({0}, device)));
+    }
+  }
+
+  std::vector<c10::IValue*> outputsVec;
+  outputsVec.reserve(outputsVec_.size());
+  for (auto& output : outputsVec_) {
+    outputsVec.push_back(&output);
+  }
+
+  Caffe2Operator(schema, std::move(inputsVec), std::move(outputsVec)).Run();
+
+  torch::jit::drop(*stack, total_num_arguments);
+  for (auto& output: outputsVec_) {
+    torch::jit::push(*stack, std::move(output));
+  }
+
+  // postcondition: All inputs are cleared from the stack, there's now one
+  //                IValue for each output which holds the result. This
+  //                might reuse one of the preallocated tensors but doesn't have to.
+}
+
+inline c10::FunctionSchema make_function_schema_for_c10(const char* OperatorName, std::vector<c10::Argument> inputs, std::vector<c10::Argument> outputs) {
+  // actual_inputs is the real inputs plus an optional tensor argument for each output.
+  // this can be used to pass in a preallocated output tensor.
+  std::vector<c10::Argument> actual_inputs = inputs;
```

Comment:
You already made a copy of `inputs` when this function got called right? Doing another copy seems a bit pointless

Reply:
yes, should be moved instead. good catch.